src/Microsoft.ML.Core/Data/ReadOnlyMemoryUtils.cs

```diff
@@ -0,0 +1,464 @@
+﻿using Microsoft.ML.Runtime;
+using Microsoft.ML.Runtime.Internal.Utilities;
+using System;
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+using System.Text;
+
+namespace Microsoft.ML.Runtime.Data
+{
+    public static class ReadOnlyMemoryUtils
+    {
+
+        /// <summary>
+        /// This method retrieves the raw buffer information. The only characters that should be
+        /// referenced in the returned string are those between the returned min and lim indices.
+        /// If this is an NA value, the min will be zero and the lim will be -1. For either an
+        /// empty or NA value, the returned string may be null.
+        /// </summary>
+        public static string GetRawUnderlyingBufferInfo(out int ichMin, out int ichLim, ReadOnlyMemory<char> memory)
+        {
+            MemoryMarshal.TryGetString(memory, out string outerBuffer, out ichMin, out int length);
+            ichLim = ichMin + length;
+            return outerBuffer;
+        }
+
+        public static int GetHashCode(this ReadOnlyMemory<char> memory) => (int)Hash(42, memory);
+
+        public static bool Equals(this ReadOnlyMemory<char> memory, object obj)
+        {
+            if (obj is ReadOnlyMemory<char>)
+                return Equals((ReadOnlyMemory<char>)obj, memory);
+            return false;
+        }
+
+        /// <summary>
+        /// This implements IEquatable's Equals method. Returns true if both are NA.
+        /// For NA propagating equality comparison, use the == operator.
+        /// </summary>
+        public static bool Equals(ReadOnlyMemory<char> b, ReadOnlyMemory<char> memory)
+        {
+            if (memory.Length != b.Length)
```

Comment:
This method and the one below could probably efficiently enough be done over direct indices over the `ReadOnlyMemory<char>` structure, at least, so I would hope. This would simplify the method considerably.